isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java

```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
   public RelNode visit(Project project) throws RuntimeException {
     RelNode child = project.getInput().accept(this);
     List<RexNode> rexList =
-        project.getExpressions().stream()
+        project.getExpressions().stream().map(expr -> expr.accept(expressionRexConverter)).toList();
+
+    return relBuilder.push(child).project(rexList).build();
+  }
+
+  @Override
+  public RelNode visit(Join join) throws RuntimeException {
+    var left = join.getLeft().accept(this);
+    var right = join.getRight().accept(this);
+    var condition =
+        join.getCondition()
+            .map(c -> c.accept(expressionRexConverter))
+            .orElse(relBuilder.literal(true));
+    var joinType =
+        switch (join.getJoinType()) {
+          case INNER -> JoinRelType.INNER;
+          case LEFT -> JoinRelType.LEFT;
+          case RIGHT -> JoinRelType.RIGHT;
+          case OUTER -> JoinRelType.FULL;
+          case SEMI -> JoinRelType.SEMI;
+          case ANTI -> JoinRelType.ANTI;
+          case UNKNOWN -> throw new UnsupportedOperationException(
+              "Unknown join type is not supported");
+        };
+    return relBuilder.push(left).push(right).join(joinType, condition).build();
+  }
+
+  @Override
+  public RelNode visit(Aggregate aggregate) throws RuntimeException {
+    RelNode child = aggregate.getInput().accept(this);
+    var groupExprLists =
+        aggregate.getGroupings().stream()
             .map(
-                expr ->
-                    expr.accept(
-                        new ExpressionRexConverter(
-                            relOptCluster.getTypeFactory(), scalarFunctionConverter)))
+                gr ->
+                    gr.getExpressions().stream()
+                        .map(expr -> expr.accept(expressionRexConverter))
+                        .toList())
             .toList();
+    List<RexNode> groupExprs =
+        groupExprLists.stream().flatMap(Collection::stream).collect(Collectors.toList());
+    RelBuilder.GroupKey groupKey = relBuilder.groupKey(groupExprs, groupExprLists);
 
-    return relBuilder.push(child).project(rexList).build();
+    List<AggregateCall> aggregateCalls =
+        aggregate.getMeasures().stream().map(m -> fromMeasure(m)).toList();
```

Comment:
For this kind of thing, using method handle `.map(this::fromMeasure).toList()` is more readable.

Reply:
changed.